pkg/controller/s3/bucket.go

```diff
@@ -0,0 +1,199 @@
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube        client.Client
+	newClientFn func(ctx context.Context, credentials []byte, region string, auth awsclients.AuthMethod) (s3.BucketClient, error)
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1beta1.Bucket)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	p := &awsv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderReference.Name}, p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if aws.BoolValue(p.Spec.UseServiceAccount) {
+		bucketClient, err := c.newClientFn(ctx, []byte{}, p.Spec.Region, awsclients.UsePodServiceAccount)
+		return &external{s3client: bucketClient, kube: c.kube, clients: bucketclients.MakeClients(cr, bucketClient)}, errors.Wrap(err, errCreateBucketClient)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	bucketClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{s3client: bucketClient, kube: c.kube, clients: bucketclients.MakeClients(cr, bucketClient)}, errors.Wrap(err, errCreateBucketClient)
+}
+
+type external struct {
+	kube     client.Client
+	s3client s3.BucketClient
+	clients  []bucketclients.BucketResource
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.Bucket)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+	if _, err := e.s3client.HeadBucketRequest(&awss3.HeadBucketInput{Bucket: aws.String(meta.GetExternalName(cr))}).Send(ctx); err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(s3.IsNotFound, err), errHead)
+	}
+	for _, client := range e.clients {
+		updated, err := client.Observe(ctx, cr)
+		if err != nil {
+			return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: false}, err
+		}
+		if updated == bucketclients.NeedsUpdate {
+			return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: false}, nil
+		}
+		if updated == bucketclients.NeedsDeletion {
+			return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: false}, nil
+		}
```

Comment:
```suggestion
		status, err := client.Observe(ctx, cr)
		if err != nil {
			return managed.ExternalObservation{}, err
		}
		if status == bucketclients.NeedsUpdate || status == bucketclients.NeedsDeletion {
			return managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: false}, nil
		}
```

I believe this would correspond to the same result.